### PR TITLE
proxmox_snap: allow set retention for snapshots

### LINKED ISSUE #6576: community.general.proxmox_snap trim snapshots
### Summary

I would like to be able to trim proxmox snapshots. e.g. I have a gitlab pipeline that snapshots a lxc before each deployment. Over time this will hurt performance and storage. I would like to have the option to create a new snapshot and only keep the last X snapshots.

### Issue Type

Feature Idea

### Component Name

proxmox_snap

### Additional Information

```yaml (paste below)
- name: Create new container snapshot and keep only the 2 newest snapshots
  community.general.proxmox_snap:
    api_user: root@pam
    api_password: 1q2w3e
    api_host: node1
    vmid: 100
    state: present
    snapname: snapshot-42
    trim: 2
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by lexxxel**
##### SUMMARY

Add a trim option to delete older snapshots

Fixes #6576

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

proxmox_snap

##### ADDITIONAL INFORMATION

```paste below
output stays the same
```

### CODE REVIEW

File: changelogs/fragments/6576-proxmox-snap-allow-trimming.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - "add trim parameter to delete old snapshots (https://github.com/ansible-collections/community.general/pull/6576)."
```
Comment (felixfontein) [inline changelogs/fragments/6576-proxmox-snap-allow-trimming.yml:3]:
```suggestion
  - "proxmox_snap - add ``trim`` parameter to delete old snapshots (https://github.com/ansible-collections/community.general/pull/6576)."
```

File: plugins/modules/proxmox_snap.py [hunk 2]
```diff
@@ -74,6 +74,12 @@
       - Name of the snapshot that has to be created/deleted/restored.
     default: 'ansible_snap'
     type: str
+  trim:
+    description:
+      - Trim snapshots if there are more than C(trim) snapshots.
+      - If C(trim) is set to 0, all snapshots will be kept.
```
Comment (felixfontein) [inline plugins/modules/proxmox_snap.py:80]:
```suggestion
      - Trim snapshots if there are more than O(trim) snapshots.
      - If O(trim) is set to V(0), all snapshots will be kept.
```

File: plugins/modules/proxmox_snap.py [hunk 3]
```diff
@@ -74,6 +74,12 @@
...
+    default: 0
+    type: int
```
Comment (felixfontein) [inline plugins/modules/proxmox_snap.py:82]:
```suggestion
    type: int
    version_added: 7.1.0
```

File: plugins/modules/proxmox_snap.py [hunk 4]
```diff
@@ -228,7 +250,9 @@ def snapshot_create(self, vm, vmid, timeout, snapname, description, vmstate, unb
             timeout -= 1
         if vm['type'] == 'lxc' and unbind is True and mountpoints:
             self._container_mp_restore(vm, vmid, timeout, unbind, mountpoints, vmstatus)
-        return False
+
+        self.snapshot_trim(vm, vmid, trim)
```
Comment (felixfontein) [inline plugins/modules/proxmox_snap.py:254]: Right now the trimming is only done when a snapshot is really created. If the snapshot to be created already exists, no trimming is done.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution. I've added some first comments.

File: plugins/modules/proxmox_snap.py (hunk 2 above)
Comment (UnderGreen) [reply plugins/modules/proxmox_snap.py:80]: I would call that param `retention` as it is a widespread name for that type of operation.

Comment (lexxxel) [reply plugins/modules/proxmox_snap.py:80]: That is a great suggestion :bow:. It is obviously the right name for the feature.

File: plugins/modules/proxmox_snap.py (hunk 4 above)
Comment (lexxxel) [reply plugins/modules/proxmox_snap.py:255]: True. This is like I intended it. However, I'm open for discussions. In theory, If we do NOT create a new snapshot, we should NOT have to discard snapshots.

**Review by UnderGreen — APPROVED** [review]

File: plugins/modules/proxmox_snap.py [hunk 5]
```diff
@@ -74,6 +74,13 @@
       - Name of the snapshot that has to be created/deleted/restored.
     default: 'ansible_snap'
     type: str
+  retention:
+    description:
+      - remove old snapshots if there are more than O(retention) snapshots.
```
Comment (felixfontein) [inline plugins/modules/proxmox_snap.py:79]:
```suggestion
      - Remove old snapshots if there are more than O(retention) snapshots.
```

File: plugins/modules/proxmox_snap.py [hunk 6]
```diff
@@ -74,6 +74,13 @@
...
+      - If O(retention) is set to V(0), all snapshots will be kept.
```
Comment (felixfontein) [inline plugins/modules/proxmox_snap.py:80]: The behavior should be properly documented:
```suggestion
      - If O(retention) is set to V(0), all snapshots will be kept.
      - This is only used when O(state=present) and when an actual snapshot is created.
        If no snapshot is created, all existing snapshots will be kept.
```

File: plugins/modules/proxmox_snap.py (hunk 4 above)
Comment (felixfontein) [reply plugins/modules/proxmox_snap.py]: There's always the use-case of cleaning up ansible-managed snapshots at another point of time than creation. Say you want to create them with ansible during some time-critical operation and want to do the cleanup later. Right now it's not possible (without creating yet another snapshot).

In the end I'm OK with both behaviors, as long as it is documented.

**Review by UnderGreen — APPROVED** [review]

**Comment by felixfontein** [comment]
@lexxxel thanks for your contribution!
@UnderGreen thanks for reviewing!

**Pull request merged**